examples/Example21/example21.cpp

```diff
@@ -0,0 +1,127 @@
+#include "FTFP_BERT_AdePT.hh"
+#include "G4HadronicProcessStore.hh"
+#include "G4EmParameters.hh"
+#include "G4VisExecutive.hh"
+#include "G4UIExecutive.hh"
+#include <sstream>
+
+#include "AdePTTrackingManager.hh"
+
+int main(int argc, char **argv)
+{
+  // Macro name from arguments
+  G4String batchMacroName;
+  G4String outputDirectory  = "";
+  G4String outputFilename   = "";
+  bool doBenchmark          = false;
+  bool doValidation         = false;
+  G4bool useInteractiveMode = true;
+  G4bool useAdePT           = true;
+  G4String helpMsg("Usage: " + G4String(argv[0]) +
+                   " [option(s)] \n No additional arguments triggers an interactive mode "
+                   "executing vis.mac macro. \n Options:\n\t-h\t\tdisplay this help "
+                   "message\n\t-m MACRO\ttriggers a batch mode executing MACRO\n");
+  for (G4int i = 1; i < argc; ++i) {
+    G4String argument(argv[i]);
+    if (argument == "-h" || argument == "--help") {
+      G4cout << helpMsg << G4endl;
+      return 0;
+    } else if (argument == "-m") {
+      batchMacroName     = G4String(argv[i + 1]);
+      useInteractiveMode = false;
+      ++i;
+    } else if (argument == "--output_dir") {
+      outputDirectory = G4String(argv[i + 1]);
+      ++i;
+    } else if (argument == "--output_file") {
+      outputFilename = G4String(argv[i + 1]);
+      ++i;
+    } else if (argument == "--do_benchmark") {
+      doBenchmark = true;
+    } else if (argument == "--do_validation") {
+      doValidation = true;    
+    } else if (argument == "--no_AdePT") {
+      useAdePT = false;
+    } else {
+      G4Exception("main", "Unknown argument", FatalErrorInArgument,
+                  ("Unknown argument passed to " + G4String(argv[0]) + " : " + argument + "\n" + helpMsg).c_str());
+    }
+  }
+
+#if defined TEST
+  if (doBenchmark && doValidation) {
+    G4Exception(
+        "main()", "Notification", JustWarning,
+        "The options --do_benchmark and --do_validation are mutually exclusive! --do_benchmark will be ignored");
+  }
+  if (!doBenchmark && !doValidation) {
+    G4Exception(
+        "main()", "Notification", JustWarning,
+        "Testing is enabled but no option has been selected, data will not be collected for this run.\n"
+        "Available options are:\n"
+        "--do_benchmark\n"
+        "--do_validation");
+  }
+#else
+  if (doBenchmark || doValidation) {
+    G4Exception(
+        "main()", "Notification", JustWarning,
+        "The application must be compiled with -DTEST in order to use the options --do_benchmark and --do_validation");
+  }
+#endif  
+
+  // Initialization of default Run manager
+  auto *runManager = G4RunManagerFactory::CreateRunManager(G4RunManagerType::Default);
+  //auto *runManager = G4RunManagerFactory::CreateRunManager(G4RunManagerType::Serial);
+
+ // Physics list
+ //
+ AdePTTrackingManager* trackingManager = nullptr;
+
+ if(useAdePT){
+  trackingManager = new AdePTTrackingManager;
```

Comment:
The example currently only works with one thread. The reason is that we need one tracking manager per thread (the same way we had one fast simulation model per thread)...

Reply:
Argh... I new it was too easy... ;-) OK, I won't be able to look into it this week. If anybody has some time to touch this code, please go ahead. Otherwise, I am back next week...